Marlin/src/pins/pins_ARCHIM2.h

```diff
@@ -203,7 +203,7 @@
 
   #define SD_DETECT_PIN     2   // D2  PB25_TIOA0
 
-  #if ENABLED(NEWPANEL)
+  #if ENABLED(NEWPANEL) || ENABLED(EXTENSIBLE_UI)
```

Comment:
I guess the same question applies to the above
```cpp
#if ENABLED(ULTRA_LCD) || ENABLED(EXTENSIBLE_UI)
```
Either there's a real display with these real pins, or there's not. It's unclear how `EXTENSIBLE_UI` causes these pins to exist, if the `ULTRA_LCD` display itself doesn't have them.

Reply:
@thinkyhead: The only reason I add those definitions is that we have been using the pins from the RepRapDiscount connectors for our display. We use all the pins on the SC2 10 pin header, so this corresponds to the NEWPANEL. Most boards have the RepRapDiscount SC1 and SC2, so it makes sense for us to reuse those pins.

There's nothing about EXTENSIBLE_UI that dictates which pins are used, but reusing the pins from the RepRapDiscount display is a great starting point since those connectors are generally available and connected to GPIO pins.

Reply:
For what it is worth, the boards we will be producing will use a 10-wire ribbon cable just like the one the RepRapDiscount uses, but give different meanings to the pins in the EXP2 connector (the EXP1 connector will not be used). This is why we want all the RepRapDiscount pins defined... so we can use them in our implementation of the EXTENSIBLE_UI.

Reply:
@thinkyhead: I think the reason this gets confusing is that Marlin currently labels pins based on their ULTRALCD/NEWPANEL functionality, but we are using (misusing?) those pins and the associated connector for another different display. Since those pins are already defined for all the boards we use (Archim 2, Mini-Rambo, Rambo, EinsyRetro) it made perfect sense for us to do this, since it allows folks to replace their old RepRapDiscount displays with our new Color touch panels and it will work with just a change of firmware.

Reply:
You have the right general idea. Marlin's pins files assign pin numbers to specific functions. You wouldn't want to make code that uses these names but then assigns different functions to them. Instead, as you suggest, you would want to make it clear where the pins are physically located on the specific board, such as the EXPn or AUXn headers, and then your extensible code could decide how to use those labeled pins.

Reply:
> You wouldn't want to make code that uses these names but then assigns different functions to them.

@thinkyhead: Guilty as charged, sort of. I am doing that right now, but I reassign the names to other variables that are correctly named with regards to their functions, so it's only a little bit evil, not completely evil:

`#define CLCD_SOFT_SPI_SCLK             BTN_ENC`

Not perfect, but at least you won't find code SPI code using BTN_ENC directly. I think these shenanigans simply point to the fact that refactoring the "pins.h" files would be helpful. BTN_ENC isn't a pin name, it's a function assigned to that pin by UltraLCD. A better pin name would be EXP1_PIN2, if we adopt the RepRapDiscount connector as a de facto standard (which seems reasonable, since I suspect a good number of boards out there support it already)

Breaking things out like this across all "pins.h" files would probably needs to be done with some sort of automated script. I wouldn't take the chance to do it by hand and introduce mistakes. Maybe a clever scripting project for some day.

Reply:
If someone wants to examine each board and add these alternative labels for pins according to where they appear in AUX or EXP headers, that would be fine. There's no automated way to do this. You'll have to download the schematics or examine/test the boards physically to determine which pin numbers belong to each breakout header.

Reply:
> There's no automated way to do this. You'll have to download the schematics or examine/test the boards physically to determine which pin numbers belong to each breakout header.

@thinkyhead: Are you sure about this? I was thinking that for every board that currently supports the RepRapDiscount display, it would be possible to deduce pin names automatically by working backwards from the RepRapDiscount function names. For example, if a board defines LCD_PINS_RS, then we know from the RepRapDiscount display schematics that pin would be called EXP1_PIN4. Not all boards may have those pins grouped together in one header labeled EXP1, but at least we know that GPIO pin connects to EXP1_PIN4 on the display side.

Reply:
Working backwards from the UltraLCD definitions would allow us to name about 12 unique GPIO pins from EXP1 and EXP2, plus the hardware SCLK, MOSI and MISO pins that are present on EXP2.  This should be sufficient for most EXTENSIBLE_UIs to function.